### PR TITLE
Shelf block crud

### PR DESCRIPTION
**Pull request by siripr4**
Change log:
- `shelf_block` CRUD operations and tests
- Inline test table in handler tests

### CODE REVIEW

File: warehouse-management-service/pkg/database/postgres/shelf_block.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+)
+
+type shelfBlockQueries interface {
+	createShelfBlockTx(ctx context.Context, tx *sql.Tx, block wms.ShelfBlock) error
+	getShelfBlockByIdTx(ctx context.Context, tx *sql.Tx, id string) (wms.ShelfBlock, error)
+	updateShelfBlockTx(ctx context.Context, tx *sql.Tx, block wms.ShelfBlock) error
+	deleteShelfBlockTx(ctx context.Context, tx *sql.Tx, id string) error
+	warehouseExistsTx(ctx context.Context, tx *sql.Tx, id string) (bool, error)
+}
+
```
Comment (shhivam) [inline warehouse-management-service/pkg/database/postgres/shelf_block.go:17]: Should we add the go-mockgen command for this?

File: warehouse-management-service/pkg/database/postgres/shelf_block_test.go [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+}
+
+func TestGetShelfBlockByIdTxError(t *testing.T) {
+	tx, err := shelfService.db.Begin()
+	if err != nil {
+		t.Error(err)
+	}
+	defer tx.Rollback()
+
+	_, err = shelfService.queries.getShelfBlockByIdTx(context.Background(), tx, "bad_id")
```
Comment (shhivam) [inline warehouse-management-service/pkg/database/postgres/shelf_block_test.go:83]: Should we assert on the actual error also and not just checking if it's nil or not? What do you think?

File: warehouse-management-service/pkg/database/postgres/warehouse.go [hunk 3]
```diff
@@ -9,7 +9,7 @@ import (
 	_ "github.com/lib/pq"
 )
 
-type queries interface {
+type warehouseQueries interface {
```
Comment (shhivam) [inline warehouse-management-service/pkg/database/postgres/warehouse.go:12]: Same for here. Should we add a go-mockgen command for this interface?

File: warehouse-management-service/pkg/database/postgres/shelf_block.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+type shelfBlockQueriesImpl struct{}
+
+type ShelfBlockService struct {
+	queries shelfBlockQueries
+	db      *sql.DB
+}
+
+var InvalidWarehouse = errors.New("invalid warehouseId")
```
Comment (shhivam) [inline warehouse-management-service/pkg/database/postgres/shelf_block.go:25]: Let's not make a static variable for an error. Ideally, there should be a new call to `errors.New` whenever we want to return an error. 

From the [docs](https://pkg.go.dev/errors#New):
> Each call to New returns a distinct error value even if the text is identical

**Review by shhivam — CHANGES_REQUESTED** [review]

File: warehouse-management-service/pkg/database/postgres/shelf_block.go (hunk 4 above)
Comment (shhivam) [reply warehouse-management-service/pkg/database/postgres/shelf_block.go:25]: An `error` can have other meta information which will be lost if we just use a static definition.

File: warehouse-management-service/pkg/database/postgres/shelf_block_mock.go
[the remark this answers is not in the document]
Comment (siripr4) [reply warehouse-management-service/pkg/database/postgres/shelf_block_mock.go:1]: Yes, the mock files need to be in the `mock` subdirectory. I've refactored this in a later branch.

File: warehouse-management-service/pkg/database/postgres/shelf_block.go (hunk 1 above)
Comment (siripr4) [reply warehouse-management-service/pkg/database/postgres/shelf_block.go:17]: Added, missed this.

File: warehouse-management-service/pkg/database/postgres/shelf_block_test.go (hunk 2 above)
Comment (siripr4) [reply warehouse-management-service/pkg/database/postgres/shelf_block_test.go:83]: The intent here is to check if `getShelfBlockByIdTx` returns an error when the underlying query returns an error. This is why I considered validating the error itself to be unnecessary (not intended to test the library).

One thing I can think of why checking for the exact error makes sense is to ensure that the error is returned as is without any abstraction. This is what is intended here and if we want to do so in the future, the tests will have to be updated as well making the robust.

Comment (siripr4) [reply warehouse-management-service/pkg/database/postgres/shelf_block_test.go:83]: Changed this to assert on the actual error

File: warehouse-management-service/pkg/database/postgres/warehouse.go (hunk 3 above)
Comment (siripr4) [reply warehouse-management-service/pkg/database/postgres/warehouse.go:12]: Added

File: warehouse-management-service/pkg/database/postgres/shelf_block.go (hunk 4 above)
Comment (siripr4) [reply warehouse-management-service/pkg/database/postgres/shelf_block.go:25]: This error is returned when the validation of a foreign key relation fails, it does not replace any underlying error (check [usage](https://github.com/nilenso/siris-onboarding/pull/31/files#diff-d40dcbfb80511563d7f2e94f43c6376b948ffbb424e9edb560244c085a652f2fR112))

Moreover, having package-level error vars is a common pattern to define custom errors, as seen in the [stdlib](https://github.com/golang/go/blob/master/src/io/io.go) too. 

However, in a later refactor, I am planning on removing this error abstraction and instead parsing the SQL library error itself.

Comment (shhivam) [reply warehouse-management-service/pkg/database/postgres/shelf_block.go:26]: >Moreover, having package-level error vars is a common pattern to define custom errors, as seen in the [stdlib](https://github.com/golang/go/blob/master/src/io/io.go) too.

Nice. TIL!

**Pull request merged**